### ISSUE TITLE
Total Loss and IOU is zero for second class of custom dataset of two

### ISSUE THREAD

**Issue by Kanyade** [AlexeyAB/darknet#7970]
Greetings!

I would like to train a custom detector with two classes but the training shows:
![image](https://user-images.githubusercontent.com/55270082/128243858-550450b4-105d-4ebc-92a3-47cefba01f66.png)
I am training on Google Colab to use their GPU.

After training I can detect the class 0 (logically the first class) with any input, and the second class works as well, until I feed it an image which it was trained on. Detecting from random images does not work with the second class.

What could be the reason behind it?

Config file:

[net]
# Testing
batch=24
subdivisions=8
# Training
# batch=64
# subdivisions=2
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 500200
policy=steps
steps=400000,450000
scales=.1,.1

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

###########

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=21
activation=linear

[yolo]
mask = 3,4,5
anchors= 1.02,1.61, 1.91,2.84, 3.54,3.94, 3.79,7.63, 6.11,10.27, 10.06,11.61
classes=2
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 8

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=21
activation=linear

[yolo]
mask = 0,1,2
anchors = 1.02,1.61, 1.91,2.84, 3.54,3.94, 3.79,7.63, 6.11,10.27, 10.06,11.61
classes=2
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

**Comment by akashAD98** [comment]
what is the size of your dataset? have you changed cfg parameters? how many steps you have trained

**Comment by Kanyade** [comment]
Roughly 500 images, 300 for first class, 200 for second training starts
with high total loss for both classes, the second being significantly
higher(3x). The IOU is still zero for the second at start as well.
The total loss changes to zero somewhere in the first 1-2000 training cycle.

In the config I changed the anchors, the batch number from 64 to 24,
subdivisions from 2 to 4, class numbers to 2, and filters to 21 before
starting the training.


Akash A Desai ***@***.***> ezt írta (időpont: 2021. aug. 5.,
Csü 7:56):

**Comment by akashAD98** [comment]
have you read this ? [https://github.com/AlexeyAB/darknet#how-to-improve-object-detection](url)
I think your dataset is not enough to learn the things, if it's possible add more images in the dataset.Alexyab suggested that  2k images /class. Also, try with default anchors. & you need to train for at least 6k steps because you have 2 classes. you will get good results.

**Comment by Kanyade** [comment]
Your link is outdated or I don't know what you are refering to, it just guides to the main issues page. 

I trained them for 6-7k steps and the mAP seems good on the test files for BOTH of the classes (./darknet detector map ....) but I still feel like something is off as it shouldn't display these values and it should recognize the second class from random pictures. 

The dataset is fine as I know, but maybe I'll do it all over again with an other tool ( I did it with the main.py in here https://github.com/rafiuddinkhan/Yolo-Training-GoogleColab/tree/master/data_converstion/main )

**Comment by stephanecharette** [comment]
You don't have to do it all over again.  But at least *verify* using another tool to ensure it is done correctly.  I recommend DarkMark, I wrote it specifically for this purpose:  https://www.ccoderun.ca/darkmark/Summary.html#DarkMarkReview  It also exports Darknet project files, which can also be useful to ensure your .cfg and .data files are setup correctly:  https://www.ccoderun.ca/darkmark/Summary.html#DarkMarkDarknet

**Comment by Kanyade** [comment]
Thank you, will look into it.

I tried it tonight after you mentioned it, but already got errors on my pc
with the darknet make command, so i will possibly deal with it next week
and check back to this issue with the dataset's status.

Stéphane Charette ***@***.***> ezt írta (időpont: 2021. aug.
5., Csü 21:13):

**Comment by Kanyade** [comment]
As I get to understand yolo better,  I realised that it's completely
normal. I am training on a tiny yolo config and as you can see in the
config I copied or in the original, it does not use layers above the first
16. (Route fields).

And the lack of detection for the second class may come from the fact that
this class can be very differently looking and that it only detects from
certain images, but with good mAP,  means that I will have to separate it
to multiple classes according to its model or to feed it more training
materials.

Stéphane Charette ***@***.***> ezt írta (időpont: 2021. aug.
5., Csü 21:13):

**Comment by haviduck** [comment]
can you paste the corresponding txt file of a confirmed correct labelled image and just a bit of the train.txt?

**Comment by Kanyade** [comment]
I am on mobile, hope it shows well from email, but here you are

haviduck ***@***.***> ezt írta (időpont: 2021. aug. 7., Szo
2:36):

**Comment by Kanyade** [comment]
It does not post images from email as I see but will post it from pc on
Monday.

haviduck ***@***.***> ezt írta (időpont: 2021. aug. 7., Szo
2:36):
